Don't suggest the use of  `impl Trait` in closure parameter

rperier opened the pull request:
Fixes #138932

rperier committed:
Suggest the use of `impl Trait` in function parameter only

Currently in case of a Trait object in closure parameter, the compiler
suggests either to use a reference, which is correct or to use an
`impl Trait` which is not. Do not emit this suggestion when the parameter
is part of a closure.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14091280606/job/39468370447) [(plain)](https://github.com/rust-lang/rust/commit/133718c52e6f04e4dd9afb2bc9c1895b75fdcb4c/checks/39468370447/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.960 Building wheels for collected packages: reuse
#12 2.961   Building wheel for reuse (pyproject.toml): started
#12 3.173   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.174   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.174   Stored in directory: /tmp/pip-ephem-wheel-cache-0yzd_p79/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.176 Successfully built reuse
#12 3.177 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.585 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.585 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.134 Collecting virtualenv
#12 4.171   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.229      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 77.0 MB/s eta 0:00:00
#12 4.285 Collecting platformdirs<5,>=3.9.1
#12 4.288   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.325 Collecting filelock<4,>=3.12.2
#12 4.338   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.355 Collecting distlib<1,>=0.3.7
#12 4.359   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.366      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 89.3 MB/s eta 0:00:00
#12 4.447 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.632 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.29.3
#12 4.632 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.7s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      137152 kB
DirectMap2M:     7202816 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.224
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 5928 files
tidy check
tidy error: file `tests/ui/traits/issue-138932.rs` must begin with a descriptive name, consider `{reason}-issue-138932.rs`
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:47
  local time: Wed Mar 26 19:05:10 UTC 2025
  network time: Wed, 26 Mar 2025 19:05:10 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rperier commented:
r? rust-lang/diagnostics

mejrs commented on tests/ui/traits/issue-138932.rs:
```diff
@@ -0,0 +1,4 @@
+fn main() {
```
If you want to reference the issue, a comment like this is the way to go
```
// issue 138932, suggestion to use impl trait in closure parameter is invalid
```

mejrs commented on compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:
```diff
@@ -3018,12 +3018,16 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
-                                Applicability::MaybeIncorrect,
-                            );
+                            if let Some(hir_id) = hir_id
+                                && self.tcx.parent_hir_node(hir_id).ident().is_some()
```
`.ident().is_some()` includes (and does not include) a lot of things. it's better for clarity and exhaustiveness to do something like this:

`!matches!(node, Node::Expr(Expr { kind: ExprKind::Closure(_), .. }))`

or (better) to enumerate where this suggestion *is* valid:

`matches!(node, /* everywhere impl Trait argument is valid */)`

rperier replied:
Indeed ! I have tried to use the second (better) approach and the right type

mejrs commented on compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs:
```diff
@@ -3018,12 +3018,16 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
-                                "impl ",
-                                Applicability::MaybeIncorrect,
-                            );
+                            if let Some(hir_id) = hir_id
```
please add a comment like "don't recommend impl Trait as a closure argument"

mejrs reviewed (commented):
Thanks! Regarding the naming of the file, please don't name it `issue-###` , but something descriptive like `dont-suggest-impl-as-closure-arg.rs`.

rperier commented:
Fixed and rebased onto master. Do I need to split into two commits ? One for the tests and one for the fix ? Regarding the git history some people do it some other don't. The maintainer has the last word :)  . What do you prefer ?

mejrs commented:
> Do I need to split into two commits ? One for the tests and one for the fix ?

that's just my personal preference, it is not necessary

Noratrieb commented:
@bors r+

bors commented:
:pushpin: Commit 8b6ff4a378e2df03db108bcd918a77f8a759334f has been approved by `Noratrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Noratrieb 8b6ff4a378e2df03db108bcd918a77f8a759334f -->
<!-- homu: {"type":"Approved","sha":"8b6ff4a378e2df03db108bcd918a77f8a759334f","approver":"Noratrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
